File: packages/npm/lib/review/detectors_struct.js
// SPDX-License-Identifier: MIT
'use strict';
/**
 * Structural detectors. Mirrors agent_style.review.detectors_struct in Python.
 * Implements RULE-A, RULE-C, RULE-E; RULE-02, 07, 09, 10 stub as "skipped".
 */

function run(rule, text, filePath) {
  const fn = DISPATCH[rule.id];
  if (!fn) {
    return {
      rule: rule.id,
      severity: rule.severity,
      detector: 'structural',
      status: 'skipped',
      count: 0,
      note: `structural detector for ${rule.id} is deferred to a future release`,
      violations: [],
    };
  }
  const violations = fn(text);
  return {
    rule: rule.id,
    severity: rule.severity,
    detector: 'structural',
    status: violations.length > 0 ? 'violation' : 'ok',
    count: violations.length,
    note: '',
    violations,
  };
}

function excerpt(textSlice, span, width = 120) {
  const [s, e] = span;
  let start = Math.max(0, s - 10);
  let end = Math.min(textSlice.length, start + width);
  let out = textSlice.slice(start, end).trim();
  if (start > 0) out = '\u2026' + out;
  if (end < textSlice.length) out = out + '\u2026';
  return out;
}

function fenceMask(text) {
  const lines = text.split(/\r\n|[\n\r]/);
  let inside = false;
  const mask = [];
  for (const line of lines) {
    if (/^\s*(```|~~~)/.test(line)) {
      inside = !inside;
      mask.push(true);
      continue;
    }
    mask.push(inside);
  }
  return mask;
}

function paragraphs(text) {
  const lines = text.split(/\r\n|[\n\r]/);
  const fence = fenceMask(text);
  const out = [];
  let curStart = null;
  let curLines = [];
  for (let i = 0; i < lines.length; i++) {
    const lineNo = i + 1;
    const line = lines[i];
    if (fence[i] || /^\s*[|#>]/.test(line)) {
      if (curLines.length > 0) {
        out.push({ start: curStart, lines: curLines });
        curStart = null;
        curLines = [];
      }
      continue;
    }
    if (!line.trim()) {
      if (curLines.length > 0) {
        out.push({ start: curStart, lines: curLines });
        curStart = null;
        curLines = [];
      }
      continue;
    }
    if (curStart === null) curStart = lineNo;
    curLines.push(line);
  }
  if (curLines.length > 0) out.push({ start: curStart, lines: curLines });
  return out;
}

function sentencesInParagraph(lines) {
  const joined = lines.map((l) => l.trim()).join(' ');
  const raw = joined.split(/(?<=[.!?])\s+(?=[A-Z"'(\[])/);
  return raw.map((s) => s.trim()).filter((s) => s.length > 0);
}

function firstWord(sentence) {
  const m = sentence.replace(/^\s+/, '').match(/[A-Za-z][A-Za-z'-]*/);
  return m ? m[0].toLowerCase() : '';
}

// ---------- RULE-A bullet overuse ------------------------------------------

// Groups: (1) leading indentation, (2) marker, (3) item content. The indent is
// captured so a nested list is not collected into its parent's group.
const BULLET_RE = /^(\s*)([*+-]|\d+\.)\s+(.*)$/;

// Width of `indent` in columns, expanding tabs to the next tab stop. Comparing
// raw character length would call one space and one tab the same depth, and
// four spaces and one tab different depths, so a list's nesting would depend on
// which whitespace the author typed.
function indentColumns(indent, tabStop = 4) {
  let width = 0;
  for (const ch of indent) {
    width += ch === '\t' ? tabStop - (width % tabStop) : 1;
  }
  return width;
}

// Coordinators, subordinators and relative pronouns. An item opening with one
// of these points back at something the bullet was severed from. Bare
// prepositions are deliberately absent: `For Linux, use apt` heads a genuine
// per-platform enumeration, so treating `for`/`to`/`with`/`from` as evidence
// flags exactly the lists RULE-A's directive permits.
const CONNECTIVE_OPENERS = new Set(`
  and but or nor so yet because therefore thus hence however although though
  while whereas since unless until then also moreover furthermore
  additionally rather which who whom whose that
`.trim().split(/\s+/));
// A repeated template is only evidence when it is an actual clause opening.
// Keying on the first two words of anything flags `Run the migration` /
// `Run the tests`, which is the directive's own checklist case.
const CLAUSE_SUBJECTS = new Set(['i', 'we', 'you', 'he', 'she', 'it', 'they', 'this', 'that', 'these', 'those']);
const COPULAS = new Set(['am', 'are', 'is', 'was', 'were']);

// Lowercase word tokens of a list item, punctuation and emphasis stripped.
function itemWords(content) {
  return content.split(/\s+/).map((w) => w.replace(/^[.,;:!?()[\]`"'*_~]+|[.,;:!?()[\]`"'*_~]+$/g, '').toLowerCase());
}

// Name the shape that marks the items as one sentence split across bullets, as
// {strength, reason}; null when the list reads as a genuine enumeration. A
// `strong` signal is self-sufficient; a `weak` one needs every item short too.
function readsAsShreddedProse(items) {
  const tokenized = items.map((it) => itemWords(it.content).filter((w) => w));
  for (const words of tokenized) {
    // A bare `And` / `But` / `Or` is a label, not connective tissue. Only a
    // connective that actually leads into something continues a sentence.
    if (words.length > 1 && CONNECTIVE_OPENERS.has(words[0])) {
      return { strength: 'strong', reason: `an item opens with the connective '${words[0]}'` };
    }
  }
  // `Per minute` / `Per user` trailing `Free tier` are fragments continuing a
  // stem. Two boundaries keep this from swallowing genuine enumerations: a list
  // where *every* item leads with `per` is a real enumeration, and so is one
  // where only a single item does not (`Global` / `Per user` / `Per project`).
  // Fragments need at least two items to continue from.
  const perCount = tokenized.filter((words) => words.length && words[0] === 'per').length;
  if (perCount > 1 && perCount < tokenized.length - 1) {
    return { strength: 'strong', reason: "later items are repeated 'per' fragments" };
  }
  const templates = new Map();
  for (const words of tokenized) {
    if (words.length < 2 || !CLAUSE_SUBJECTS.has(words[0]) || !COPULAS.has(words[1])) continue;
    const key = `${words[0]} ${words[1]}`;
    const seen = (templates.get(key) || 0) + 1;
    templates.set(key, seen);
    if (seen >= 2) return { strength: 'weak', reason: `items repeat the clause template '${key}'` };
  }
  return null;
}

/**
 * Flag short lists that read as fragmented prose rather than enumeration.
 *
 * The marker is not evidence either way, and shortness is not evidence on its
 * own either -- a checklist of terse imperative steps is short *and* genuine.
 * The test is RULE-A's own directive: "bullets are sentence shards with
 * connective tissue stripped", graded by how much each shape proves alone. A
 * strong signal (an opening coordinator/subordinator/relative pronoun, or some
 * but not all items leading with `Per`) fires regardless of item length. A weak
 * one (a repeated subject-and-copula opening) needs every item short too.
 * Imperative checklist steps, preposition-led enumerations and short
 * independent labels match neither and stay clean.
 */
function ruleA(text) {
  const out = [];
  const lines = text.split(/\r\n|[\n\r]/);
  const fence = fenceMask(text);
  let i = 0;
  while (i < lines.length) {
    if (fence[i]) { i++; continue; }
    const head = BULLET_RE.exec(lines[i]);
    if (!head) { i++; continue; }
    // Collect the siblings of this bullet. Ignoring indentation flattens a
    // nested list into its parent, so `- Rate limits:` plus two indented
    // `- Per …` children reads as one three-item group that no author wrote.
    // Deeper lines are skipped rather than ending the group, so a parent list
    // that resumes after a nested block stays one list.
    const groupStart = i + 1;
    const groupIndent = indentColumns(head[1]);
    const items = [];
    while (i < lines.length && !fence[i]) {
      const m = BULLET_RE.exec(lines[i]);
      if (!m) break;
      const indent = indentColumns(m[1]);
      if (indent > groupIndent) { i++; continue; } // a child, not a sibling
      if (indent < groupIndent) break;
      items.push({ lineNo: i + 1, content: m[3].trim() });
      i++;
    }
    const n = items.length;
    const shortItems = items.filter((it) => it.content.split(/\s+/).length <= 8).length;
    const signal = n >= 3 ? readsAsShreddedProse(items) : null;
    if (signal !== null && (signal.strength === 'strong' || shortItems === n)) {
      out.push({
        rule: 'RULE-A',
        line: groupStart,
        column: 1,
        excerpt: excerpt(lines[groupStart - 1], [0, lines[groupStart - 1].length]),
        detail: `list has ${n} items and ${signal.reason}; consider prose`,
      });
    }
  }
  return out;
}

// ---------- RULE-C consecutive same-starts ---------------------------------

function ruleC(text) {
  const out = [];
  for (const p of paragraphs(text)) {
    const sentences = sentencesInParagraph(p.lines);
    if (sentences.length < 2) continue;
    const firsts = sentences.map((s) => firstWord(s));
    for (let i = 0; i < firsts.length - 1; i++) {
      const window = firsts.slice(i, i + 3);
      if (window.length < 2) continue;
      const counts = {};
      for (const w of window) {
        if (!w) continue;
        counts[w] = (counts[w] || 0) + 1;
      }
      const dup = Object.entries(counts).find(([w, c]) => c >= 2 && w);
      if (dup) {
        out.push({
          rule: 'RULE-C',
          line: p.start,
          column: 1,
          excerpt: excerpt(sentences[i], [0, sentences[i].length]),
          detail: `consecutive sentences start with '${dup[0]}'`,
        });
        break;
      }
    }
  }
  return out;
}

// ---------- RULE-E paragraph closers ---------------------------------------

const CLOSER_STARTERS = [
  'Overall,', 'In summary,', 'In conclusion,', 'To summarize,',
  'All in all,', 'In short,', 'Ultimately,', 'Thus,', 'Therefore,',
];
const CLOSER_STARTER_RE = new RegExp(
  '^(?:' + CLOSER_STARTERS.map((s) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')).join('|') + ')',
  'i'
);
const CLOSER_PATTERNS = [
  CLOSER_STARTER_RE,
  /\bthese (?:changes|contributions|improvements|updates|results) (?:represent|demonstrate|reflect)\b/i,
  /\ba significant step (?:forward|change|improvement)\b/i,
  /\brepresents? a (?:significant|major|substantial) (?:advance|step|improvement)\b/i,
];

function ruleE(text) {
  const out = [];
  for (const p of paragraphs(text)) {
    const sentences = sentencesInParagraph(p.lines);
    if (sentences.length === 0) continue;
    const last = sentences[sentences.length - 1];
    for (const pat of CLOSER_PATTERNS) {
      if (pat.test(last)) {
        out.push({
          rule: 'RULE-E',
          line: p.start + Math.max(0, p.lines.length - 1),
          column: 1,
          excerpt: excerpt(last, [0, last.length]),
          detail: 'paragraph ends with a summary / restatement',
        });
        break;
      }
    }
  }
  return out;
}

const DISPATCH = {
  'RULE-A': ruleA,
  'RULE-C': ruleC,
  'RULE-E': ruleE,
};

module.exports = { run };

File: packages/pypi/agent_style/review/detectors_struct.py
# SPDX-License-Identifier: MIT
"""Structural detectors: line/paragraph-level heuristics.

Rules with deterministic structural detectors:
  RULE-A  bullet overuse (lists whose items read as prose shredded into bullets)
  RULE-C  consecutive same-starts (≥ 2 of 3 adjacent sentences share opening token)
  RULE-E  paragraph-closing summaries (last sentence restates the paragraph topic)

Rules in the structural bucket that still have no detector and report
"skipped" (as of v0.4.2):
  RULE-02  passive voice when agent matters
  RULE-07  positive form for affirmative claims
  RULE-09  parallel structure across coordinate items
  RULE-10  keep related words together (subject/verb proximity)

These skipped rules return ``status="skipped"`` with a note explaining why, so
the JSON output remains schema-complete for every rule in the matrix.
"""

from __future__ import annotations

import re
from typing import Callable

from agent_style.review.loader import Rule
from agent_style.review.primitive import RuleResult, Violation


def run(rule: Rule, text: str, file_path: str) -> RuleResult:
    """Dispatch to the structural detector for ``rule.id``."""
    func = _DISPATCH.get(rule.id)
    if func is None:
        return RuleResult(
            rule=rule.id,
            severity=rule.severity,
            detector="structural",
            status="skipped",
            count=0,
            note=f"structural detector for {rule.id} is deferred to a future release",
        )
    violations = func(text)
    status = "violation" if violations else "ok"
    return RuleResult(
        rule=rule.id,
        severity=rule.severity,
        detector="structural",
        status=status,
        count=len(violations),
        violations=violations,
    )


# ---------- Helpers ---------------------------------------------------------


def _excerpt(text: str, span: tuple[int, int], width: int = 120) -> str:
    start = max(0, span[0] - 10)
    end = min(len(text), start + width)
    out = text[start:end].strip()
    if start > 0:
        out = "…" + out
    if end < len(text):
        out = out + "…"
    return out


_FENCE_OPEN_RE = re.compile(r"^\s*(```|~~~)")


def _fence_mask(text: str) -> list[bool]:
    """Return a list parallel to text.splitlines() marking lines inside fences."""
    lines = text.splitlines()
    inside = False
    mask = []
    for line in lines:
        if _FENCE_OPEN_RE.match(line):
            inside = not inside
            mask.append(True)  # fence markers themselves count as inside
            continue
        mask.append(inside)
    return mask


def _paragraphs(text: str) -> list[tuple[int, list[str]]]:
    """Yield (start_line_number, paragraph_lines) for each prose paragraph.

    Skips fenced code blocks and table rows entirely.
    """
    lines = text.splitlines()
    fence = _fence_mask(text)
    out: list[tuple[int, list[str]]] = []
    cur_start = None
    cur_lines: list[str] = []
    for i, line in enumerate(lines, start=1):
        idx = i - 1
        if fence[idx] or line.lstrip().startswith(("|", "#", ">")):
            if cur_lines:
                out.append((cur_start or 1, cur_lines))
                cur_start, cur_lines = None, []
            continue
        if not line.strip():
            if cur_lines:
                out.append((cur_start or 1, cur_lines))
                cur_start, cur_lines = None, []
            continue
        if cur_start is None:
            cur_start = i
        cur_lines.append(line)
    if cur_lines:
        out.append((cur_start or 1, cur_lines))
    return out


def _sentences_in_paragraph(lines: list[str]) -> list[str]:
    """Join paragraph lines and split into sentences."""
    joined = " ".join(l.strip() for l in lines)
    # Conservative sentence splitter: period/exclamation/question followed by space + capital.
    raw = re.split(r"(?<=[.!?])\s+(?=[A-Z\"'\(\[])", joined)
    return [s.strip() for s in raw if s.strip()]


def _first_word(sentence: str) -> str:
    m = re.match(r"[A-Za-z][A-Za-z'-]*", sentence.lstrip())
    return m.group(0).lower() if m else ""


# ---------- RULE-A bullet overuse -------------------------------------------


# Groups: (1) leading indentation, (2) marker, (3) item content. The indent is
# captured so a nested list is not collected into its parent's group.
_BULLET_RE = re.compile(r"^(\s*)([*+-]|\d+\.)\s+(.*)$")


def _indent_columns(indent: str, tab_stop: int = 4) -> int:
    """Width of ``indent`` in columns, expanding tabs to the next tab stop.

    Comparing raw character length would call one space and one tab the same
    depth, and four spaces and one tab different depths, so a list's nesting
    would depend on which whitespace the author typed.
    """
    width = 0
    for ch in indent:
        width += tab_stop - (width % tab_stop) if ch == "\t" else 1
    return width

# Coordinators, subordinators and relative pronouns. An item opening with one
# of these points back at something the bullet was severed from. Bare
# prepositions are deliberately absent: `For Linux, use apt` heads a genuine
# per-platform enumeration, so treating `for`/`to`/`with`/`from` as evidence
# flags exactly the lists RULE-A's directive permits.
_CONNECTIVE_OPENERS = frozenset(
    """
    and but or nor so yet because therefore thus hence however although though
    while whereas since unless until then also moreover furthermore
    additionally rather which who whom whose that
    """.split()
)
# A repeated template is only evidence when it is an actual clause opening.
# Keying on the first two words of anything flags `Run the migration` /
# `Run the tests`, which is the directive's own checklist case.
_CLAUSE_SUBJECTS = frozenset(
    {"i", "we", "you", "he", "she", "it", "they", "this", "that", "these", "those"}
)
_COPULAS = frozenset({"am", "are", "is", "was", "were"})


def _item_words(content: str) -> list[str]:
    """Lowercase word tokens of a list item, punctuation and emphasis stripped."""
    return [w.strip(".,;:!?()[]`\"'*_~").lower() for w in content.split()]


def _reads_as_shredded_prose(items: list[tuple[int, str]]) -> tuple[str, str] | None:
    """Name the shape that marks the items as one sentence split across bullets.

    Returns ``(strength, reason)``, or None when the list reads as a genuine
    enumeration. A ``strong`` signal is self-sufficient evidence; a ``weak`` one
    only counts when every item is also short.
    """
    tokenized = [[w for w in _item_words(content) if w] for _, content in items]
    for words in tokenized:
        # A bare `And` / `But` / `Or` is a label, not connective tissue. Only a
        # connective that actually leads into something continues a sentence.
        if len(words) > 1 and words[0] in _CONNECTIVE_OPENERS:
            return "strong", f"an item opens with the connective '{words[0]}'"
    # `Per minute` / `Per user` trailing `Free tier` are fragments continuing a
    # stem. Two boundaries keep this from swallowing genuine enumerations: a
    # list where *every* item leads with `per` is a real enumeration, and so is
    # one where only a single item does not (`Global` / `Per user` /
    # `Per project`). Fragments need at least two items to continue from.
    per_count = sum(bool(words) and words[0] == "per" for words in tokenized)
    if 1 < per_count < len(tokenized) - 1:
        return "strong", "later items are repeated 'per' fragments"
    templates: dict[tuple[str, str], int] = {}
    for words in tokenized:
        if len(words) < 2 or words[0] not in _CLAUSE_SUBJECTS or words[1] not in _COPULAS:
            continue
        key = (words[0], words[1])
        templates[key] = templates.get(key, 0) + 1
        if templates[key] >= 2:
            return "weak", f"items repeat the clause template '{key[0]} {key[1]}'"
    return None


def _rule_a(text: str) -> list[Violation]:
    """Flag short lists that read as fragmented prose rather than enumeration.

    The marker is not evidence either way: RULE-A's directive permits genuine
    enumerations (API endpoints, config options, checklist steps) without
    privileging ordered over unordered, and a numbered causal chain is exactly
    the fragmented prose the rule targets. Shortness is not evidence on its own
    either -- a checklist of terse imperative steps is short *and* genuine.

    So the test is the directive's own: "bullets are sentence shards with
    connective tissue stripped". Two shapes betray a shredded sentence, graded
    by how much they prove on their own:

    1. **Strong.** An item opens with a coordinator, subordinator or relative
       pronoun (``Because``, ``Which``, ``And``, ``Rather``), or some but not
       all items lead with ``Per``. Either is self-sufficient evidence, so it
       fires regardless of item length; an explicit ``Because`` shard is no
       less a shard for sitting in a list with one long item.
    2. **Weak.** Two or more items repeat a subject-and-copula opening
       (``It is fast`` / ``It is accurate``). Only counts when every item is
       also short.

    Bare prepositions are not evidence, and neither is an arbitrary repeated
    first-two-words key. ``For Linux, use apt`` heads a genuine per-platform
    enumeration and ``Run the migration`` / ``Run the tests`` is the
    directive's own checklist case; both stay clean, as do short independent
    labels. The cost is recall on ambiguous triads such as ``It accelerates
    retrieval`` / ``It improves accuracy``, where the second token varies.
    Those need a semantic judge, not a sharper regex.
    """
    out: list[Violation] = []
    lines = text.splitlines()
    fence = _fence_mask(text)
    i = 0
    while i < len(lines):
        if fence[i]:
            i += 1
            continue
        m = _BULLET_RE.match(lines[i])
        if not m:
            i += 1
            continue
        # Start of a bullet group; collect the siblings of this bullet. Ignoring
        # indentation flattens a nested list into its parent, so `- Rate limits:`
        # plus two indented `- Per …` children reads as one three-item group
        # that no author wrote. Deeper lines are skipped rather than ending the
        # group, so a parent list that resumes after a nested block stays one
        # list.
        group_start = i + 1  # 1-indexed line number
        group_indent = _indent_columns(m.group(1))
        items: list[tuple[int, str]] = []
        while i < len(lines) and not fence[i]:
            bullet = _BULLET_RE.match(lines[i])
            if not bullet:
                break
            indent = _indent_columns(bullet.group(1))
            if indent > group_indent:
                i += 1  # a child of the current item, not a sibling
                continue
            if indent < group_indent:
                break
            items.append((i + 1, bullet.group(3).strip()))
            i += 1
        n = len(items)
        short_items = sum(1 for _, content in items if len(content.split()) <= 8)
        signal = _reads_as_shredded_prose(items) if n >= 3 else None
        if signal is not None and (signal[0] == "strong" or short_items == n):
            reason = signal[1]
            out.append(
                Violation(
                    rule="RULE-A",
                    line=group_start,
                    column=1,
                    excerpt=_excerpt(lines[group_start - 1], (0, len(lines[group_start - 1]))),
                    detail=f"list has {n} items and {reason}; consider prose",
                )
            )
    return out


# ---------- RULE-C consecutive same-starts ----------------------------------


def _rule_c(text: str) -> list[Violation]:
    """Flag when ≥ 2 of any 3 consecutive sentences share their opening word."""
    out: list[Violation] = []
    for start_line, para_lines in _paragraphs(text):
        sentences = _sentences_in_paragraph(para_lines)
        if len(sentences) < 2:
            continue
        firsts = [_first_word(s) for s in sentences]
        for i in range(len(firsts) - 1):
            # Window of 3 (or 2 at the tail): flag if ≥ 2 match.
            window = firsts[i : i + 3]
            if len(window) < 2:
                continue
            # Count duplicates in the window.
            seen: dict[str, int] = {}
            for w in window:
                if not w:
                    continue
                seen[w] = seen.get(w, 0) + 1
            if any(count >= 2 for word, count in seen.items() if word):
                # Report the first offending sentence's starting word.
                dup_word = next(w for w, c in seen.items() if c >= 2 and w)
                out.append(
                    Violation(
                        rule="RULE-C",
                        line=start_line,
                        column=1,
                        excerpt=_excerpt(sentences[i], (0, len(sentences[i]))),
                        detail=f"consecutive sentences start with '{dup_word}'",
                    )
                )
                break  # one finding per paragraph is enough
    return out


# ---------- RULE-E paragraph-closing summaries ------------------------------


_CLOSER_STARTERS = (
    "Overall,",
    "In summary,",
    "In conclusion,",
    "To summarize,",
    "All in all,",
    "In short,",
    "Ultimately,",
    "Thus,",
    "Therefore,",
)
_CLOSER_PATTERNS = (
    re.compile(r"^(?:" + "|".join(re.escape(s) for s in _CLOSER_STARTERS) + r")", re.IGNORECASE),
    re.compile(r"\bthese (?:changes|contributions|improvements|updates|results) (?:represent|demonstrate|reflect)\b", re.IGNORECASE),
    re.compile(r"\ba significant step (?:forward|change|improvement)\b", re.IGNORECASE),
    re.compile(r"\brepresents? a (?:significant|major|substantial) (?:advance|step|improvement)\b", re.IGNORECASE),
)


def _rule_e(text: str) -> list[Violation]:
    """Flag paragraph-closing sentences that restate or summarize."""
    out: list[Violation] = []
    for start_line, para_lines in _paragraphs(text):
        sentences = _sentences_in_paragraph(para_lines)
        if not sentences:
            continue
        last = sentences[-1]
        for pat in _CLOSER_PATTERNS:
            if pat.search(last):
                # Line number is start_line + number of lines before the last
                out.append(
                    Violation(
                        rule="RULE-E",
                        line=start_line + max(0, len(para_lines) - 1),
                        column=1,
                        excerpt=_excerpt(last, (0, len(last))),
                        detail="paragraph ends with a summary / restatement",
                    )
                )
                break
    return out


# ---------- Dispatch --------------------------------------------------------


_DISPATCH: dict[str, Callable[[str], list[Violation]]] = {
    "RULE-A": _rule_a,
    "RULE-C": _rule_c,
    "RULE-E": _rule_e,
    # RULE-02, RULE-07, RULE-09, RULE-10 fall through to "skipped" in run()
}
